src/automata/include/automata/ta_product.hpp

```diff
@@ -22,64 +22,134 @@
+	// The same as above, but for the final locations.
+	std::transform(std::begin(std::begin(automata)->get_final_locations()),
+	               std::end(std::begin(automata)->get_final_locations()),
+	               std::inserter(product_final_locations, end(product_final_locations)),
+	               [](const auto &location) { return ProductLocation{{location.get()}}; });
+	std::for_each(std::next(std::begin(automata)),
+	              std::end(automata),
+	              [&product_final_locations](const auto &ta) {
+		              std::set<ProductLocation> augmented_product_final_locations;
+		              for (const auto &product_location : product_final_locations) {
+			              for (const auto &ta_location : ta.get_final_locations()) {
+				              auto new_product_location = product_location;
+				              new_product_location->push_back(ta_location);
+				              augmented_product_final_locations.insert(new_product_location);
+			              }
+		              }
+		              product_final_locations = augmented_product_final_locations;
+	              });
+
+	// Assert that each product location consists of one location from each TA.
+	assert(std::all_of(begin(product_locations),
+	                   end(product_locations),
+	                   [&automata](const auto &location) {
+		                   return location.get().size() == automata.size();
+	                   }));
+
+	std::set<ActionT>                                        product_alphabet;
+	ProductLocation                                          product_initial_location;
+	std::set<std::string>                                    product_clocks;
+	std::vector<Transition<std::vector<LocationT>, ActionT>> product_transitions;
+	for (const auto &[ta_i, ta] : ranges::views::enumerate(automata)) {
+		product_alphabet.insert(std::begin(ta.get_alphabet()), std::end(ta.get_alphabet()));
+		product_initial_location->push_back(ta.get_initial_location().get());
+		product_clocks.insert(std::begin(ta.get_clocks()), std::end(ta.get_clocks()));
+		for (const auto &[source_location, transition] : ta.get_transitions()) {
+			for (const auto &product_source_location : product_locations) {
+				if (product_source_location.get()[ta_i] == transition.source_.get()) {
+					for (const auto &product_target_location : product_locations) {
+						// All the locations must be equal except for the location at ta_i, which switches to
+						// the target location of the transition.
+						if (std::equal(begin(product_source_location.get()),
+						               begin(product_source_location.get()) + ta_i,
+						               begin(product_target_location.get()))
+						    && product_target_location.get()[ta_i] == transition.target_.get()
+
+						    && std::equal(begin(product_source_location.get()) + ta_i + 1,
+						                  end(product_source_location.get()),
+						                  begin(product_target_location.get()) + ta_i + 1)) {
```

Comment:
That is a very lengthy condition, should we not maybe compute the separate properties before, for readability, or add comments?

Reply:
The condition was unnecessarily complicated, thanks for catching this!